src/wp-includes/blocks.php

```diff
@@ -1287,22 +1303,44 @@ function insert_hooked_blocks_into_rest_response( $response, $post ) {
 			'ignoredHookedBlocks' => $ignored_hooked_blocks,
 		);
 	}
+
+	if ( 'wp_navigation' === $post->post_type ) {
+		$wrapper_block_type = 'core/navigation';
+	} else {
+		$wrapper_block_type = 'core/post-content';
+	}
+
 	$content = get_comment_delimited_block_content(
-		'core/navigation',
+		$wrapper_block_type,
 		$attributes,
 		$response->data['content']['raw']
 	);
 
-	$content = apply_block_hooks_to_content( $content, $post );
+	$content = apply_block_hooks_to_content(
+		$content,
+		$post,
+		'insert_hooked_blocks_and_set_ignored_hooked_blocks_metadata'
+	);
 
-	// Remove mock Navigation block wrapper.
+	// Remove mock block wrapper.
 	$content = remove_serialized_parent_block( $content );
 
 	$response->data['content']['raw'] = $content;
 
+	// No need to inject hooked blocks twice.
+	$priority = has_filter( 'the_content', 'apply_block_hooks_to_content' );
+	if ( false !== $priority ) {
+		remove_filter( 'the_content', 'apply_block_hooks_to_content', $priority );
+	}
+
 	/** This filter is documented in wp-includes/post-template.php */
 	$response->data['content']['rendered'] = apply_filters( 'the_content', $content );
 
+	// Add back the filter.
+	if ( false !== $priority ) {
+		add_filter( 'the_content', 'apply_block_hooks_to_content', $priority );
+	}
```

Comment:
This is all about calling `apply_block_hooks_to_content` (above) with a different third argument, is that right? It's confusing to see that it's called above then the filter that would call it here is removed, but it makes sense if it's about that third argument.

The comments could be a bit clearer.

```suggestion
	// `apply_block_hooks_to_content` is called above. Ensure it is not called again as a filter.
	$priority = has_filter( 'the_content', 'apply_block_hooks_to_content' );
	if ( false !== $priority ) {
		remove_filter( 'the_content', 'apply_block_hooks_to_content', $priority );
	}

	/** This filter is documented in wp-includes/post-template.php */
	$response->data['content']['rendered'] = apply_filters( 'the_content', $content );

	// Restore the filter if it was set initially.
	if ( false !== $priority ) {
		add_filter( 'the_content', 'apply_block_hooks_to_content', $priority );
	}
```

Reply:
> This is all about calling `apply_block_hooks_to_content` (above) with a different third argument, is that right? It's confusing to see that it's called above then the filter that would call it here is removed, but it makes sense if it's about that third argument.

Yeah, that's pretty much the reason. We're calling `apply_block_hooks_to_content` here with the `insert_hooked_blocks_and_set_ignored_hooked_blocks_metadata` callback, as we'd like the block markup returned by the REST API for use in the editor to contain both hooked blocks, _and_ the `metadata.ignoredHookedBlocks` attributes.

After that step, we apply the `the_content` filter to the blocks markup (which at this also includes hooked blocks and `ignoredHookedBlocks` metadata) to get the "rendered" version (which -- among other things -- runs `do_blocks` to expand block delimiters into "rendered" markup, by running individual blocks' `render_block` methods etc).

OTOH, we have "globally" added `apply_block_hooks_to_content` to the `the_content` filter. (This is necessary for hooked blocks to be inserted into Post Content on the frontend -- including sites that use a Classic Theme rather than a Block Theme.) By default, `apply_block_hooks_to_content` uses `insert_hooked_blocks` as its callback, which is fine for the frontend (we don't need to set `ignoredHookedBlocks` metadata there).

But since we've already applied Block Hooks earlier in this process, we now remove the filter and re-add it only after the (other) `the_content` filter have been applied.

Thinking about it now, there _might_ be alternatives, such as using `insert_hooked_blocks_and_set_ignored_hooked_blocks_metadata` for the frontend (i.e. in `the_content`) as well; but it seems kinda wasteful to set `ignoredHookedBlocks` metadata when we don't even need it. So I'm inclined to leave as-is for now, but open to reconsidering and tweaking this some more. LMK if you have any other thoughts or suggestions, maybe I'm not seeing all possible alternatives.
 
> The comments could be a bit clearer.

Those make a lot of sense, I'll apply them. Thank you!

Reply:
> Thinking about it now, there _might_ be alternatives, such as using `insert_hooked_blocks_and_set_ignored_hooked_blocks_metadata` for the frontend (i.e. in the_content) as well; but it seems kinda wasteful to set `ignoredHookedBlocks` metadata when we don't even need it

There is some appeal in removing the filter swapping implementation that's here. What's involved in the wastful processing, is it parsing the block with the HTML API unnecessarily or something along those lines?